builtin/providers/aws/resource_aws_api_gateway_authorizer.go

```diff
@@ -0,0 +1,189 @@
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceAwsApiGatewayAuthorizerCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+	// Create the gateway
+	log.Printf("[DEBUG] Creating API Gateway Authorizer")
+
+	var credentials *string
+	if v, ok := d.GetOk("credentials"); ok {
+		credentials = aws.String(v.(string))
+	}
+
+	var result_in_ttl *int64
+	if v, ok := d.GetOk("result_in_ttl"); ok {
+		result_in_ttl = aws.Int64(int64(v.(int)))
+	}
+
+	var identity_validation_expression *string
+	if v, ok := d.GetOk("identity_validation_expression"); ok {
+		identity_validation_expression = aws.String(v.(string))
+	}
+
+	var err error
+	authorizer, err := conn.CreateAuthorizer(&apigateway.CreateAuthorizerInput{
+		AuthorizerUri:  aws.String(d.Get("authorizer_uri").(string)),
+		IdentitySource: aws.String(d.Get("identity_source").(string)),
+		Name:           aws.String(d.Get("name").(string)),
+		RestApiId:      aws.String(d.Get("rest_api_id").(string)),
+		Type:           aws.String(d.Get("type").(string)),
+
+		AuthorizerCredentials:        credentials,
+		AuthorizerResultTtlInSeconds: result_in_ttl,
+		IdentityValidationExpression: identity_validation_expression,
+	})
+	if err != nil {
+		return fmt.Errorf("Error creating API Gateway Authorizer: %s", err)
+	}
+
+	d.SetId(*authorizer.Id)
+	log.Printf("[DEBUG] API Gateway Authorizer ID: %s", d.Id())
+
+	return nil
+}
+
+func resourceAwsApiGatewayAuthorizerRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+
+	log.Printf("[DEBUG] Reading API Gateway Deployment %s", d.Id())
+	out, err := conn.GetAuthorizer(&apigateway.GetAuthorizerInput{
+		RestApiId:    aws.String(d.Get("rest_api_id").(string)),
+		AuthorizerId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NotFoundException" {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+	log.Printf("[DEBUG] Received API Gateway Authorizer: %s", out)
+	d.SetId(*out.Id)
+
+	return nil
+}
+
+func resourceAwsApiGatewayAuthorizerUpdateOperations(d *schema.ResourceData) []*apigateway.PatchOperation {
+	operations := make([]*apigateway.PatchOperation, 0)
+
+	if d.HasChange("description") {
+		operations = append(operations, &apigateway.PatchOperation{
+			Op:    aws.String("replace"),
+			Path:  aws.String("/description"),
+			Value: aws.String(d.Get("description").(string)),
+		})
+	}
+
+	return operations
+}
+
+func resourceAwsApiGatewayAuthorizerUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+
+	log.Printf("[DEBUG] Updating API Gateway API Key: %s", d.Id())
+
+	_, err := conn.UpdateAuthorizer(&apigateway.UpdateAuthorizerInput{
+		AuthorizerId:    aws.String(d.Id()),
+		RestApiId:       aws.String(d.Get("rest_api_id").(string)),
+		PatchOperations: resourceAwsApiGatewayAuthorizerUpdateOperations(d),
+	})
+	if err != nil {
+		return err
```

Comment:
We tend to accompany errors coming from the API by some extra context - e.g. `Updating/Creation/Deletion failed: %s`. The reason is that many APIs don't provide that context and it's difficult for the user or anyone else to figure out why something has failed and in which stage.